Add quick help page

quangson91 opened the pull request:
Added a quick help page when no files are open.

Fixes #5990 

Technical changes:

**Detail**
Add quick-help-view beside the editor-tabs-split.
When there is no tabs -> show the quick help view.

Demo:

https://github.com/defold/defold/assets/6185205/30a1c9a4-25b7-4eec-8541-da97f1698758


---
**UPDATE**

Change logo style

![image](https://github.com/defold/defold/assets/6185205/4d0d3fd4-7ec6-44fc-a794-22f52faeb3f9)

quangson91 committed:
Add quick help page

Signed-off-by: quangson91 <quangson91@gmail.com>

quangson91 commented on editor/src/clj/editor/app_view.clj:
```diff
@@ -294,6 +296,81 @@
                                             (let [tab-panes (.getItems editor-tabs-split)]
+
+                                              ;TODO: Extract this logic to another method.
+                                              (let [total-tabs (reduce + (map #(-> ^TabPane % (.getTabs) (.size)) tab-panes))
```
@matgis @britzl 
I really want to extract this code to another function.

I have try create a function (in this file) like this:

```
(defn- update-quick-help-pane [arg]
; implement here
)
```

But I don't know how to call this method from here.
Could you please suggest?

Thx!

quangson91 commented on editor/src/clj/editor/app_view.clj:
```diff
@@ -294,6 +296,81 @@
+
+                                                    ;TODO: Do nothing if the box-items has children
+                                                    ;TODO: Move the items to the other place where we can define shortcut for each OS
+                                                    (let [items [{:label "Re-Open Closed File" :keys ["Ctrl" "Shift" "T"]}
```
I also want to extract these config.
Where is the good place to put this one?
I think we also need dynamic config (by platform). Because each platform may have different shortcut.

matgis replied:
Users can override the keymap using a config file, so we can't hardcode these.

The `AppView` node you're editing has access to the keymap. Add `keymap-config` as an input dependency to the surrounding `g/fnk`. This ensures the view will be updated if the keymap changes when we get around to making a keymap editor in the future.
```
(g/fnk [^SplitPane editor-tabs-split open-views open-dirty-views keymap-config]
  ...)
```
You can then do something like:
```
(let [command->key-combo
      (into {}
            (mapcat (fn [[key-combo command-infos]]
                      (map (fn [command-info]
                             [(:command command-info) key-combo])
                           command-infos)))
            keymap)]
  ;; You can now get the key-combo info for each command, i.e.
  (command->key-combo :reopen-recent-file)
  (command->key-combo :open-asset)
  ...)
```
The key-combo is a map that looks like this:
```
{:key "T",
 :alt-down? false,
 :control-down? false,
 :meta-down? true,
 :shift-down? true,
 :shortcut-down? false}
```
`:shortcut-down?` will be `false`, and should be ignored. On macOS, **Cmd** (`:meta-down?`) is used instead of **Ctrl** for basic shortcuts, but **Ctrl** may be used for some advanced shortcuts. You can assume `:meta-down?` will be `false` on non-mac platforms.

quangson91 replied:
I used your suggestion. For now, the key-map can be dynamic display by OS.

quangson91 commented on editor/styling/stylesheets/_editor.scss:
```diff
@@ -201,3 +201,11 @@ Links:
   -fx-text-fill: #89d2ff;
 }
+
+.key-button {
```
I tried put these config directly in the editor.css.
Then later, I copy the config here.

I would like to know, what is the process when build to generate css from scss.

quangson91 committed:
Update quick help logo

Signed-off-by: quangson91 <quangson91@gmail.com>

matgis commented on editor/resources/quick-help-view.fxml:
```diff
@@ -0,0 +1,17 @@
+<?import javafx.scene.layout.*?>
+<VBox id="quick-help-box" alignment="CENTER" spacing="10" visible="false" AnchorPane.bottomAnchor="0.0" AnchorPane.leftAnchor="0.0" AnchorPane.rightAnchor="0.0" AnchorPane.topAnchor="0.0" xmlns="http://javafx.com/javafx/17.0.2-ea" xmlns:fx="http://javafx.com/fxml/1">
+    <ImageView fitHeight="150.0" fitWidth="200.0" pickOnBounds="true" preserveRatio="true">
+        <Image url="@logo-ver-outline-dark-320.png" />
```
I think we should use the SVG version of the logo. Perhaps the white one with some transparency could work? You can get it in SVG format from here: https://defold.com/logo-and-trademark/

quangson91 replied:
Does javafx support display svg natively?

For now, I change to use the Defold logo `logo-ver-outline-white.png`, and it looks pretty good.

matgis replied:
I think it does. Did you try it?

quangson91 replied:
Yes, I already tried. It display nothing when we use svg format.

![image](https://github.com/defold/defold/assets/6185205/9ecc914c-b984-409b-b768-c2e15c0107c1)

matgis replied:
Hmm, yes. It seems it supports SVG shape syntax using the `<SVGPath>` element, but not the full SVG spec. Oh well.

aglitchman commented:
I thought my issue is one of those that will never be done. Because it is not so essential, but rather for UI beautification.

@quangson91 You are doing very well! 

P.S. I would 2x the size of the logo and increase the line spacing.

quangson91 committed:
Make quick help logo bigger

Signed-off-by: quangson91 <quangson91@gmail.com>

quangson91 committed:
Extract to update-quick-help-pane

Signed-off-by: quangson91 <quangson91@gmail.com>

quangson91 committed:
Refactor using data from keymap

Signed-off-by: quangson91 <quangson91@gmail.com>

quangson91 committed:
Increase quick-help items vgap

Signed-off-by: quangson91 <quangson91@gmail.com>

quangson91 committed:
Remove log

Signed-off-by: quangson91 <quangson91@gmail.com>

matgis reviewed (changes requested):
Nice work!

I think it works great. I found a few minor code style issues. I also tweaked the labels, and the list of actions we display. But somehow the logo shows up as opaque white now? I think we wanted it to be more subtly blended, right? It is kind of blinding right now.

I also provided a few Clojure tips & suggestions, but you don't have to do them if you disagree. 😅

quangson91 commented:
@matgis 
Thank you for the feedback. I will fix those issues.

For the logo, I missed update the style in the _editor.scss.

I will update it.

```
#quick-help-logo {
    -fx-opacity: 0.1;
}

```

quangson91 commented on editor/src/clj/editor/app_view.clj:
```diff
@@ -240,6 +242,88 @@
+                            (if (:alt-down? key-combo) "Alt")
+                            (if (:shift-down? key-combo) "Shift")
+                            (str (:key key-combo))))]
+    {:label label :keys (into [] keys)}))
```
I am learning Clojure. So really love the detail.

britzl commented on editor/resources/quick-help-view.fxml:
```diff
@@ -0,0 +1,17 @@
+<?import javafx.scene.layout.*?>
+<VBox id="quick-help-box" alignment="CENTER" spacing="10" visible="false" AnchorPane.bottomAnchor="0.0" AnchorPane.leftAnchor="0.0" AnchorPane.rightAnchor="0.0" AnchorPane.topAnchor="0.0" xmlns="http://javafx.com/javafx/17.0.2-ea">
+    <ImageView id="quick-help-logo" fitWidth="300.0" pickOnBounds="true" preserveRatio="true">
+        <Image url="@logo-ver-outline-white.png" />
```
Note that there is also a version of the logo for use blended into a dark background:

https://github.com/defold/assets-defold/blob/master/assets-defold/logo/logo-ver-outline-bright.png

quangson91 replied:
I updated the _editor.scss

quangson91 committed:
Refactor

Signed-off-by: quangson91 <quangson91@gmail.com>

quangson91 commented on editor/src/clj/editor/app_view.clj:
```diff
@@ -96,12 +96,14 @@
            [javafx.scene.image Image ImageView]
            [javafx.scene.input Clipboard ClipboardContent]
-           [javafx.scene.layout AnchorPane StackPane]
+           [javafx.scene.layout AnchorPane GridPane HBox StackPane VBox Region]
```
Yup. I removed.

quangson91 commented on editor/src/clj/editor/app_view.clj:
```diff
@@ -240,6 +242,88 @@
+        quick-help-box (.lookup (.getParent editor-tabs-split) "#quick-help-box")
+        ^GridPane box-items (.lookup (.getParent editor-tabs-split) "#quick-help-items")]
+
+    ;Only make quick help visible when there is no-tabs.
```
I got it. I am also update in other places.

quangson91 commented on editor/styling/stylesheets/_editor.scss:
```diff
@@ -201,3 +201,11 @@ Links:
   -fx-text-fill: #89d2ff;
 }
+
+.key-button {
```
Yup. I changed to `#quick-help-items > .key-button`

quangson91 commented:
@matgis 
Thank you for the very detail.
You guys in the Defold team is so kind.

I just update source code by your suggestions.
Please take a looK!

vlaaad commented:
I have a couple of notes:
For @quangson91:
- First of all, we are very grateful for your contributions ❤️
- Instead of creating strings like `"Cmd"` and `"Ctrl"` in `create-key-info`, rely on JavaFX's [KeyCombination::getDisplayText](https://openjfx.io/javadoc/14/javafx.graphics/javafx/scene/input/KeyCombination.html#getDisplayText()) method. This will give us nicely-looking shortcuts on macOS like <kbd>⌃⇧→</kbd> — that looks the same as in the menu bar.
- We cannot expect any shortcuts to be present — our users can create custom keymaps that don't have shortcuts for commands we want to show. We must check that the shortcut exists before invoking `create-key-info`

For @matgis:
I hope we will migrate the rest of the editor to [cljfx](github.com/cljfx/cljfx) eventually, and that makes me hesitant to introduce more code that uses raw JavaFX, fxml, and mutations of JavaFX view graph. On the other hand, I understand that this is another layer of complexity that might be too much for external contributors. Do you think we are okay with supporting the added raw JavaFX/FXML/mutability code?

matgis commented:
@vlaaad Yes, I'm fine with maintaining the raw JavaFX calls for this sort of thing. It is a simple view. We can redo it in [cljfx](https://github.com/defold/defold/pull/github.com/cljfx/cljfx) in a future PR.

I think we have to accept the fact that we currently have a mix of raw JavaFX and cljfx views, so contributors will look at what is already there when they contribute.

matgis reviewed (changes requested):
@vlaaad has a good point about the possibility of key commands not being bound if the user has a custom keymap.

You could use [keep](https://clojuredocs.org/clojure.core/keep) when you generate the list of `items` to only include commands that have a shortcut assigned. Something like this:

```clj
(keep (fn [[command label]]
        (when-some [key-combo (command->key-combo command)]
          (create-key-info label key-combo)))
      [[:open-asset "Open Asset"]
       [:reopen-recent-file "Re-Open Closed File"]
       ,,,])
```

Also, could you please add a newline between `(defn- update-quick-help-pane ...)` and `(g/defnode AppView ...)`? 🙂

quangson91 committed:
Only keep available command

Signed-off-by: quangson91 <quangson91@gmail.com>

vlaaad commented:
`key-combo->display-text` uses the same method of the KeyCombination class that I suggest using directly. You have access to KeyCombination instances in `update-quick-help-pane`: the keymap contains them in a `:key-combo` values:
```clj
(g/node-value (dev/app-view) :keymap)
=> {{:key "Up"
     :alt-down? false
     :control-down? false
     :meta-down? true
     :shift-down? true
     :shortcut-down? false} [{:command :select-beginning-of-file
                              :shortcut "Shift+Meta+Up"
                              :key-combo #object[javafx.scene.input.KeyCodeCombination 0x5d8a6b6f "Shift+Meta+Up"]}]
    ...}
```
In this output it's shown as `#object[javafx.scene.input.KeyCodeCombination 0x5d8a6b6f "Shift+Meta+Up"]`.

You need to modify how you create `command->key-combo` so that it has instances of KeyCombination as values.

quangson91 committed:
Use KeyCodeCombination#getDisplayText for display shortcut text

Signed-off-by: quangson91 <quangson91@gmail.com>

quangson91 commented:
@vlaaad  I updated in the last commit: https://github.com/defold/defold/pull/8085/commits/62041b37c99fb361e4fcb3704f9b6d75508f55e9

Could you please take a look?

The small problem is `getDisplayText` itself return the `String`. So it really hard to create a separate keys.
I mean, it returns strings like: "Ctrl+Shift+A", but it would better if we can get list of ("Ctrl", "Shift", "A").
So we can represents in the UI better.

Split by "+" sometimes work on linux/windows. But doesn't work on macOS.

So if you have any suggestion please tell me.

For  now, the UI is look like this one:
![image](https://github.com/defold/defold/assets/6185205/658b6c65-8e51-48ae-823d-67e89ac16e9a)

vlaaad commented:
Hi @quangson91!

I'm sorry that I asked you for changes so many times, now I see that my suggestion to use `.getDisplayText` made the view worse, and it looked better with how you implemented it earlier, especially on macOS where proper view would additionally require us to use a different font:
<img width="515" alt="Screenshot 2023-10-06 at 09 38 31" src="https://github.com/defold/defold/assets/732554/63307b9a-e3a1-4da4-b139-5cdc72f2cded">

Would you kindly revert your last commit? Then I think we will be happy to merge it!

quangson91 committed:
Make keys display better

Signed-off-by: quangson91 <quangson91@gmail.com>

quangson91 commented:
Hi @vlaaad 
I just updated.
The commit is revert & borrow some code from displayText.
That mean, the display key is different between mac & non-mac os.

Here is the result (I just fake is-macos = false to test; but please check on your mac, since the key using the default system-font).

![Screenshot_20231006_152925](https://github.com/defold/defold/assets/6185205/bcecc8fa-8635-4ef3-b227-87c65b66caa8)
![Screenshot_20231006_152954](https://github.com/defold/defold/assets/6185205/87d4508b-c311-4873-95da-d7b2f4759f04)

If there is thing that need to fix, please tell me.

Thanks!

matgis commented:
Well done! ❤️

It looks good on my Mac. I'm happy to merge this in its current state.

<img width="361" alt="image" src="https://github.com/defold/defold/assets/11677496/243c4687-083c-4074-a161-31794318bcce">

vlaaad commented:
Yes, looks great! Thank you @quangson91! Merge it @matgis 👍

matgis commented:
Merged! Thanks for contributing to Defold! ❤️

vlaaad commented:
@matgis We should also cherry-pick it into editor-dev I think

matgis commented:
I agree! Cherry-picked into `editor-dev`.

mikatuo commented:
@quangson91 I opened Defold editor recently and could not understand why it looked so much nicer than usual and if the logo with keys was there before. Thank you!
